source/fe/fe_face.cc

```diff
@@ -263,6 +263,158 @@ FE_FaceQ<dim,spacedim>::hp_constraints_are_implemented () const
+{
+  // this element is continuous only for the highest dimensional bounding object
+  if (dim !=2)
+    return
+      std::vector<std::pair<unsigned int, unsigned int> > ();
+  else
+    {
+      // this is similar to the FE_Q_Base class
+      if (const FE_FaceQ<dim,spacedim> *fe_q_other
+          = dynamic_cast<const FE_FaceQ<dim,spacedim>*>(&fe_other))
+        {
+          // dofs are located along lines, so two dofs are identical if they are
+          // located at identical positions.
+          // Therefore, read the points in unit_support_points for the
+          // first coordinate direction. We take the lexicographic ordering of the
+          // points in the second direction (i.e., y-direction) since we know
+          // that the first p+1 dofs are located at the left (x=0) face.
+          const unsigned int p = this->degree;
+          const unsigned int q = fe_q_other->degree;
+
+          std::vector<std::pair<unsigned int, unsigned int> > identities;
+
+          const std::vector<unsigned int> &index_map_inverse=
+            this->poly_space.get_numbering_inverse();
+          const std::vector<unsigned int> &index_map_inverse_other=
+            fe_q_other->poly_space.get_numbering_inverse();
+
+          for (unsigned int i=0; i<p+1; ++i)
+            for (unsigned int j=0; j<q+1; ++j)
+              if (std::fabs(this->unit_support_points[index_map_inverse[i]][dim-1]-
+                            fe_q_other->unit_support_points[index_map_inverse_other[j]][dim-1])
+                  < 1e-14)
+                identities.emplace_back (i, j);
+
+          return identities;
+        }
+      else if (dynamic_cast<const FE_Nothing<dim>*>(&fe_other) != nullptr)
+        {
+          // the FE_Nothing has no degrees of freedom, so there are no
+          // equivalencies to be recorded
+          return std::vector<std::pair<unsigned int, unsigned int> > ();
+        }
+      else if (fe_other.dofs_per_face == 0)
+        {
+          // if the other element has no elements on faces at all,
+          // then it would be impossible to enforce any kind of
+          // continuity even if we knew exactly what kind of element
+          // we have -- simply because the other element declares
+          // that it is discontinuous because it has no DoFs on
+          // its faces. in that case, just state that we have no
+          // constraints to declare
+          return std::vector<std::pair<unsigned int, unsigned int> > ();
+        }
+      else
+        {
+          Assert (false, ExcNotImplemented());
+          return std::vector<std::pair<unsigned int, unsigned int> > ();
+        }
+    }
+}
+
+
+
+template <int dim, int spacedim>
+std::vector<std::pair<unsigned int, unsigned int> >
+FE_FaceQ<dim, spacedim>::
+hp_quad_dof_identities (const FiniteElement<dim, spacedim> &fe_other) const
+{
+  // this element is continuous only for the highest dimensional bounding object
+  if (dim !=3)
```

Comment:
Same here -- maybe write it as `if (dim > 3)`.

I *think* that this function can not actually be called if `dim==1` or `dim==2`. But we can verify that by putting an `Assert (dim >= 3, ExcInternalError())` at the top. This then also explains that the `else` part below here really just means `dim==3`.